InterfacePython/example_optimization_02.py

```diff
@@ -28,12 +28,103 @@
+
+    # Create a surrogate model for the quadratic.
+    # NOTE: local polynomial grids are not well suited for first-order methods like gradient descent.
+    grid = SG.makeGlobalGrid(2, 1, 2, "iptotal", "leja")
+    needed_points = grid.getNeededPoints()
+    needed_values = np.resize(np.apply_along_axis(func, 1, needed_points), [grid.getNumNeeded(), 1])
+    grid.loadNeededValues(needed_values)
+    surrogate_func = lambda x : grid.evaluate(x)
+    surrogate_grad = lambda x : grid.differentiate(x)
+
+    # Run the GD algorithm on the surrogate and check the output.
+    state = Opt.GradientDescentState(x0, 0.0)
+    dInfo = Opt.GradientDescent(surrogate_grad, 1/8.0, iNumIterations, fTolerance, state)
+
+    aSolution = state.getX()
+
+    sResult = ""
+    sErrors = ""
+    for i in range(2):
+        sResult = "{0:1s}{1:13.5f}".format(sResult, aSolution[i])
+        sErrors = "{0:1s}{1:13.5e}".format(sErrors, np.abs(aSolution[i] - aGlobal[i]))
+    print("\nUsing the Gradient Descent algorithm on the SURROGATE objective function, the computed solution is:")
+    print(" iterations: {0:14}".format(dInfo['performed_iterations']))
+    print("   computed: {0:1s}".format(sResult))
+    print("      error: {0:1s}".format(sErrors))
+
+
+    # Run the adaptive GD algorithm and check the output. Note that the adaptive stepsize in the state IS used here.
+    state = Opt.GradientDescentState(x0, 1.0)
+    dInfo = Opt.AdaptiveGradientDescent(func, grad, 1.25, 1.25, iNumIterations, fTolerance, state)
+
+    aSolution = state.getX()
+
+    sResult = ""
+    sErrors = ""
+    for i in range(2):
+        sResult = "{0:1s}{1:13.5f}".format(sResult, aSolution[i])
+        sErrors = "{0:1s}{1:13.5e}".format(sErrors, np.abs(aSolution[i] - aGlobal[i]))
+    print("\nUsing the ADAPTIVE Gradient Descent algorithm on the EXACT objective function, the computed solution is:")
+    print(" iterations: {0:14}".format(dInfo['performed_iterations']))
+    print("   computed: {0:1s}".format(sResult))
+    print("      error: {0:1s}".format(sErrors))
+
+    # Run the projected adaptive GD algorithm and check the output. The domain constraint is implicitly enforced by the
+    # projection function `proj`.
+    state = Opt.GradientDescentState(x0, 1.0)
+    proj = lambda x : np.array([min(0.5, max(0.0, x[0])),
+                                min(1.5, max(0.0, x[1]))]) # projection onto the box [0.0, 0.5] x [0.0, 1.5]
+    dInfo = Opt.AdaptiveProjectedGradientDescent(func, grad, proj, 1.25, 1.25, iNumIterations, fTolerance, state)
+
+    aGlobal = np.array([0.5, 1.5])
+    aSolution = state.getX()
+
+    sResult = ""
+    sErrors = ""
+    for i in range(2):
+        sResult = "{0:1s}{1:13.5f}".format(sResult, aSolution[i])
+        sErrors = "{0:1s}{1:13.5e}".format(sErrors, np.abs(aSolution[i] - aGlobal[i]))
+    print("\nUsing the ADAPTIVE Projected Gradient Descent algorithm on the EXACT objective function, the computed solution is:")
```

Comment:
What's the difference between this and line `102`, is it the use of `proj`.

The text should mention the use of the `proj`, if that's the point of the example.

Reply:
The name of the algorithm is already stated as
```
Adaptive Projected Gradient Descent
```
compared to 
```
Adaptive Gradient Descent
```
in line `102`. Is there something else that I should add to emphasize the projection?

Reply:
You capitalize `EXACT` vs. `SURROGATE` and `ADAPTIVE` vs. non-adaptive. However, `Projected` is easily missed in the mix.

Capitalize `PROJECTED` too.

Reply:
Done. I capitalized `PROJECTED`.